---
title: "Deconvolution - OHDA model ST dataset"
author: "Yuvarani Masarapu"
date: "2023-05-26"
output: 
  html_document:
    self_contained: true
    highlight: tango
    df_print: paged
    code_folding: hide
    toc: yes
    toc_depth: 3
    toc_float:
      collapsed: false
      smooth_scroll: true    
---

# Load Required packages
```{r}
library(Seurat)
library(readr)
library(dplyr)
library(tibble)
library(EnDecon)
library(pryr)
library(reticulate)
library(philentropy)
library(Giotto)
library(spacexr)
```

# Custom functions 

## To round off dataframe values to particular digits
```{r}
round_df <- function(df, digits) {
  nums <- vapply(df, is.numeric, FUN.VALUE = logical(1))

  df[,nums] <- round(df[,nums], digits = digits)

  (df)
}
```

## Rearrange plot grids
```{r}
small.leg <- theme(legend.text = element_text(size=6), legend.key.width = unit(0.1,"cm"))
cc = scale_color_gradientn(colors = c("grey", "yellow", "red", "black"))
```

# Run deconvolution

* We use the deconvolution package EnDecon here (reference: https://academic.oup.com/bioinformatics/article/39/1/btac805/6900924)

* 14 deconvolution softwares are benchmarked - cell2location, DestVI, DSTG, Giotto/Hypergeometric, Giotto/PAGE, Giotto/rank, MIA, RCTD, Seurat, spatialDecon, spatialDWLS, SPOTlight, STdeconvolve, stereoscope, STRIDE and Tangram.

* cell2location, RCTD, spatialDWLS were best among the 14 deconvolution softwares (memory, time efficiency, celltypes detection tested etc parameters tested).

* EnDecon is a package containing combined algorithms of above top 3 softwares. It was shown to have better performance (memory and celltypes detection) when same datasets were used.

* EnDecon has some issues on macOS. I checked their original code after numerous failed tests and discovered that the code points to lots of lines written to work on a windows machine hence I was unable to run it on macOS.

* However, I was able to run deconvolution with **RCTD** (https://pubmed.ncbi.nlm.nih.gov/33603203/) which is among the top 3 deconvolution softwares and works without issues on macOS. It also has less run time compared to stereoscope.

* Another **note** - Stereoscope installation is causing a lot of issues on the server side, hence I aborted it. Upon some research, I found EnDecon paper where very nice comparisons were done.

## Read the scRNA-seq data

* Taken from https://www.nature.com/articles/nn.4216)
* Might not be the most appropriate here but used as a test to setup a deconvolution pipeline that we can use later on on relevant snRNA/scRNA-seq datasets.
* Please **note** that I used a different dataset than the one I asked for feedback over email. That dataset had data format incompatibilities which was taking too much time to solve, hence inorder to save time and build a pipeline first, I used this dataset instead.

```{r eval=FALSE}
# read scRNA-seq data
indir <- getwd()
indir

sc_data <- readRDS(paste(indir, "/brain_sc.rds", sep = ""))
sc_data@meta.data$Cell_class <- gsub("/",".",sc_data@meta.data$subclass)

# Run this if the counts are not integers (RNA assay will be unless updated after normalization)
#sc_data@assays$RNA@counts <- round(sc_data@assays$RNA@counts)
```

## Load the seurat object from the ST OHDA dataset

```{r eval=FALSE}
st_data <- readRDS("~/Documents/parkinsons_project/analysis-parkinsons/data.integrated.clusters_2022-08-22.rds")

# to see how many genes are there between the datasets
# intersect(rownames(st_data), rownames(sc_data))

#st_data@images$coordinates = data.frame(x=st_data@images$anterior1@coordinates$row,
#                                        y=st_data@images$anterior1@coordinates$col)
```

## Subset the sc data for the genes that match with the st dataset
```{r eval=FALSE}
genes.intersect <- intersect(rownames(st_data), rownames(sc_data))

sc_data <- sc_data[genes.intersect, ]
```

## Run RCTD deconvolution
```{r eval=FALSE}
#Sys.setenv(RETICULATE_PYTHON = "/Users/yuvarani.masarapu/miniconda3/envs/cell2loc_env/bin/python3")
#reticulate::use_condaenv(condaenv = "cell2loc_env" , conda = "auto", required = NULL)

use_python(python = "/Users/yuvarani.masarapu/miniconda3/envs/cell2loc_env/bin/python")

source_python('./EnDecon/cell2location_main.py') # Had to install psutil by pip install psutil in conda environment from terminal first
```

* Save image coordinates for each sample separately as it's needed later on during deconvolution. Keep it in the same order as the order of the samples being subsetted in the next section.
```{r eval=FALSE}
# Vector memory exhausted because the dataset with all samples is too big for analysis.

images <- c("slice1", "slice1.1", "slice1.2", "slice1.3", "slice1.6", "slice1.7", "slice1.12", "slice1.13", "slice1.18", "slice1.19", "slice1.20", "slice1.21",
            "slice1.4","slice1.5","slice1.8","slice1.9","slice1.10","slice1.11", "slice1.14", "slice1.15", "slice1.16", "slice1.17", "slice1.22", "slice1.23")

img_coord <- list()

# all OHDA samples first
img_coord[[1]] <- data.frame(x = st_data@images$slice1@coordinates$row,
                             y = st_data@images$slice1@coordinates$col)
img_coord[[2]] <- data.frame(x = st_data@images$slice1.1@coordinates$row,
                             y = st_data@images$slice1.1@coordinates$col)
img_coord[[3]] <- data.frame(x = st_data@images$slice1.2@coordinates$row,
                             y = st_data@images$slice1.2@coordinates$col)
img_coord[[4]] <- data.frame(x = st_data@images$slice1.3@coordinates$row,
                             y = st_data@images$slice1.3@coordinates$col)
img_coord[[5]] <- data.frame(x = st_data@images$slice1.6@coordinates$row,
                             y = st_data@images$slice1.6@coordinates$col)
img_coord[[6]] <- data.frame(x = st_data@images$slice1.7@coordinates$row,
                             y = st_data@images$slice1.7@coordinates$col)
img_coord[[7]] <- data.frame(x = st_data@images$slice1.12@coordinates$row,
                             y = st_data@images$slice1.12@coordinates$col)
img_coord[[8]] <- data.frame(x = st_data@images$slice1.13@coordinates$row,
                             y = st_data@images$slice1.13@coordinates$col)
img_coord[[9]] <- data.frame(x = st_data@images$slice1.18@coordinates$row,
                             y = st_data@images$slice1.18@coordinates$col)
img_coord[[10]] <- data.frame(x = st_data@images$slice1.19@coordinates$row,
                             y = st_data@images$slice1.19@coordinates$col)
img_coord[[11]] <- data.frame(x = st_data@images$slice1.20@coordinates$row,
                             y = st_data@images$slice1.20@coordinates$col)
img_coord[[12]] <- data.frame(x = st_data@images$slice1.21@coordinates$row,
                             y = st_data@images$slice1.21@coordinates$col)

# all controls next
img_coord[[13]] <- data.frame(x = st_data@images$slice1.4@coordinates$row,
                             y = st_data@images$slice1.4@coordinates$col)
img_coord[[14]] <- data.frame(x = st_data@images$slice1.5@coordinates$row,
                             y = st_data@images$slice1.5@coordinates$col)
img_coord[[15]] <- data.frame(x = st_data@images$slice1.8@coordinates$row,
                             y = st_data@images$slice1.8@coordinates$col)
img_coord[[16]] <- data.frame(x = st_data@images$slice1.9@coordinates$row,
                             y = st_data@images$slice1.9@coordinates$col)
img_coord[[17]] <- data.frame(x = st_data@images$slice1.10@coordinates$row,
                             y = st_data@images$slice1.10@coordinates$col)
img_coord[[18]] <- data.frame(x = st_data@images$slice1.11@coordinates$row,
                             y = st_data@images$slice1.11@coordinates$col)
img_coord[[19]] <- data.frame(x = st_data@images$slice1.14@coordinates$row,
                             y = st_data@images$slice1.14@coordinates$col)
img_coord[[20]] <- data.frame(x = st_data@images$slice1.15@coordinates$row,
                             y = st_data@images$slice1.15@coordinates$col)
img_coord[[21]] <- data.frame(x = st_data@images$slice1.16@coordinates$row,
                             y = st_data@images$slice1.16@coordinates$col)
img_coord[[22]] <- data.frame(x = st_data@images$slice1.17@coordinates$row,
                             y = st_data@images$slice1.17@coordinates$col)
img_coord[[23]] <- data.frame(x = st_data@images$slice1.22@coordinates$row,
                             y = st_data@images$slice1.22@coordinates$col)
img_coord[[24]] <- data.frame(x = st_data@images$slice1.23@coordinates$row,
                             y = st_data@images$slice1.23@coordinates$col)

rm(images)
```

* We divide the dataset into each sample object and run RCTD. This is because the bigger dataset was too big for my RAM to handle and I kept getting 'Vector memory exhausted' error. 

```{r eval=FALSE}
sample_name <- c("V11D06-099_A1","V11D06-099_C1","V11D08-305_C1","V11D08-306_A1","V11D06-069_D1","V11D06-100_D1","V11D06-070_A1","V11D06-070_B1","V11D06-069_C1","V11D06-100_B1","V11D08-305_D1","V11D08-306_B1",
                 "V11D06-100_A1","V11D08-305_A1","V11D06-069_A1","V11D06-070_C1","V11D06-099_B1","V11D06-099_D1","V11D06-069_B1","V11D06-070_D1","V11D06-100_C1","V11D08-305_B1","V11D08-306_C1","V11D08-306_D1")

st_data.list <- list()
for(obj in 1:length(sample_name)){
  st_data.list[[obj]] <- st_data[,colnames(st_data)[st_data$sample_id %in% sample_name[obj]]]
}

rm(st_data)
rm(sample_name)
```

* Now we run RCTD on the sample objects' list
```{r eval=FALSE}
results.list <- list()
for(i in 1:length(st_data.list)){
  st_data.list[[i]]@images$coordinates <- img_coord[[i]]
  results.list[[i]] <- EnDecon::RCTD_pipeline(sc_data = sc_data, st_data = st_data.list[[i]])
}

saveRDS(results.list, file = "RCTD_results.rds")
```

### Add deconvolution results to the seurat object, we save them in the "rctd_celltypes" assay
```{r eval=FALSE}
all = Reduce(rbind, results.list)
all = t(all)
all = as.data.frame(all)
all = all[ ,colnames(st_data)]
st_data[["rctd_celltypes"]] <- CreateAssayObject(counts = all)

#Save a seurat object with all the deconv data
savefile = file.path(indir, "sobject_deconv_RCTD_2023-05-27.Rds")
saveRDS(st_data, file = savefile)
```

## Deconvolution results as plots

### Celltype annotations and abbreviations used from the sc data

* Transcriptomics classes under Glutamatergic cells - L6 CT, L6b, L& IT, L2.3 IT, L4, L5 PT and L5 IT
* Inhibitory neurons class samatostatin - Sst
* Inhibitory neurons class vasoactive intestinal polypeptide - Vip
* Inhibitory neurons class parvalbumin - Pvalb
* Non-neuronal Astrocytes - Astro
* Endothelial - Endo
* GABAergic cells - Sncg
* NP - Neuropeptides (?)

### Predicted proportion of each SN-cluster onto ST-clusters 

#### Violin plots

```{r eval=FALSE}
types <- c("rctd_celltypes")
celltypes = rownames(all)
coldef_ST_H <- c("#F18F01", "#878787", "#FCCDE5", "#048BA8", "#2E4057", "#99C24D", "#B95F89", "#5F070C" ,  "#DFC27D", "#DB6C79", "#B191FF", "#157A6E", "#73683B", "#97C8EB","#C51B7D", "#BA9763", "#31081F", "#52D1DC", "#700353", "#F5D3C8", "#725752", "#D8315B", "#6B4E71", "#8CBFCF" , "#C2B2B4", "#EDE5A6", "#52B788", "#EE1B26", "#F2F230", "#91F291", "#386CB0", "#E7298A")
pie(rep(1,length(coldef_ST_H)), col = coldef_ST_H)
```


```{r, fig.height=7, fig.width=8, results='asis'}
for (ct in celltypes){
  #cat("For",type, "in", ct ,"\n")
  p = VlnPlot(st_data, features = ct, assay = "rctd_celltypes", pt.size = 0, cols = coldef_ST_H)
  #ggsave(p, filename = paste(indir,"/plots/individual-celltypes_prop/violin_", ct ,"-proportions.pdf", sep = ""), dpi =300, height=5, width=10)
  print(p)
}
```


#### Bar plots

* **Note** - We took only those SN-clusters which have atleast 10% expression in a spot.

```{r eval=FALSE}
cell.prop.cl = list()
clusters = levels(st_data$seurat_clusters)

for (type in types){
  
  cell.prop = list()
  for (cl in clusters) {
      cp = rowSums(st_data@assays[[type]]@counts[ ,st_data$seurat_clusters == cl])
      cp = cp/sum(cp)
      cp[cp < 0.1] <- 0 #cell proportions below 10% get no value
      cell.prop[[cl]] = cp
  }
  
  cell.prop = Reduce(cbind, cell.prop)
  colnames(cell.prop) = clusters
  cell.prop.cl[[type]] = cell.prop
}
```

```{r, eval=FALSE}
for (type in types){
  tot.prop = data.frame(cell.prop.cl[[type]])
  tot.prop$celltype = factor(rownames(tot.prop), levels = celltypes)

  l = reshape2::melt(tot.prop, id.vars = "celltype")

  p5 = ggplot(l, aes(x = variable, y = value, fill = celltype)) + 
    geom_bar(position = "fill", stat = "identity") + 
    RotatedAxis() + 
    scale_fill_manual(values = coldef_ST_H) + 
    ggtitle(paste(type, " in ST clusters (OHDA model dataset)", sep = "")) +
    theme(plot.title = element_text(size = 40, hjust = 0.5),
          axis.title.x = element_text(size = 30),
          axis.title.y = element_text(size = 30),
          axis.text.x = element_text(size = 20),
          axis.text.y = element_text(size = 20),
          legend.key.size = unit(1, "cm"),
          legend.title = element_text(size = 25),
          legend.text = element_text(size = 16))
  
  ggsave(p5, filename = paste(indir,"/plots/sn-clusters_per_ST-cluster/", type, "-sn-proportions.pdf", sep = ""), dpi =300, height=16, width=16)
  
  print(p5)
}
```

```{r, fig.height=8, fig.width=8}
for (type in types){
  tot.prop = data.frame(cell.prop.cl[[type]])
  tot.prop$celltype = factor(rownames(tot.prop), levels = celltypes)

  l = reshape2::melt(tot.prop, id.vars = "celltype")

  p5 = ggplot(l, aes(x = variable, y = value, fill = celltype)) + 
    geom_bar(position = "fill", stat = "identity") + 
    RotatedAxis() + 
    scale_fill_manual(values = coldef_ST_H) + 
    ggtitle(paste(type, " in ST clusters (OHDA model dataset)", sep = "")) +
    theme(plot.title = element_text(size = 20),
          axis.title.x = element_text(size = 18),
          axis.title.y = element_text(size = 18),
          axis.text.x = element_text(size = 10),
          axis.text.y = element_text(size = 10),
          legend.key.size = unit(1, "cm"),
          legend.title = element_text(size = 18),
          legend.text = element_text(size = 12))
  
  # ggsave(p5, filename = paste(indir,"/plots/sn-clusters_per_ST-cluster/", type, "-sn-proportions.pdf", sep = ""), dpi =300, height=16, width=16)
  
  print(p5)
}
```

##### Save celltype proportions for each ST cluster in a table
```{r eval=FALSE}
library("xlsx")

write.xlsx(cell.prop.cl[[types]], 
           file = paste(indir, "/plots/celltype_proportions-RCTD_onST-clusters.xlsx", sep = ""),
           sheetName=types, 
           col.names=TRUE, row.names=TRUE, append=TRUE)
```

#### Bar graphs

* Plot barplot for each celltype proportion across all ST clusters (different visualisation)

```{r eval=FALSE}
library(data.table)
library(Matrix)
library(ggplot2)

tot.prop = data.frame(cell.prop.cl[[types]])
tot.prop$celltype = factor(rownames(tot.prop), levels = celltypes)
  
celltypes <- rownames(tot.prop)
#round off proportion values to 3 decimal points
tot.prop <- round_df(tot.prop, 3)

for(ct in celltypes){
  df <- tot.prop[ct, ]
  df <- rbind(df, colnames(df))
  df <- df[, -c(33)] #remove the last column with celltype names
  df <- t(df) #transpose to arrange dataframe for plotting
  df <- data.frame(df)
  df$ST_clusters <- df$X2
  df <- df[, -c(2)]
  
  # replaced ct with 1 because spaces in ct changes the column names
  df[ ,1] <- as.numeric(df[ ,1])
  df$ST_clusters <- factor(df$ST_clusters, levels = unique(df$ST_clusters))
  
  p <- ggplot(df, aes(x = ST_clusters, y = df[,1], fill = ST_clusters)) + 
  geom_bar(position = "dodge", stat = "identity") +
  geom_text(aes(label = df[,1]), vjust=-0.5, color="black", size=6) +
  scale_fill_manual(values = coldef_ST_H) +
  labs(x="ST cluster", y="Expression", title = paste("Celltype proportions of ", ct, " in ST clusters", sep = "")) +
  theme_minimal() +
  theme(plot.title = element_text(hjust = 0.5, face = "bold", size = 25),
        axis.line = element_line(colour = "black"),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        axis.title.x = element_text(size = 25),
        axis.title.y = element_text(size = 25),
        axis.text.x = element_text(size = 20, angle = -45),
        axis.text.y = element_text(size = 20),
        legend.title = element_text(size = 20),
        legend.text = element_text(size = 15))
  
  #ggsave(filename = paste(indir, "/plots/each_celltype-proportions/", ct, ".pdf", sep = ""), plot = p, dpi = 300, height = 12, width = 14)
  
  print(p)
  
}

```

```{r, fig.height=10, fig.width=8, warning=FALSE}
library(data.table)
library(Matrix)
library(ggplot2)

tot.prop = data.frame(cell.prop.cl[[types]])
tot.prop$celltype = factor(rownames(tot.prop), levels = celltypes)
  
celltypes <- rownames(tot.prop)
#round off proportion values to 3 decimal points
tot.prop <- round_df(tot.prop, 3)

for(ct in celltypes){
  df <- tot.prop[ct, ]
  df <- rbind(df, colnames(df))
  df <- df[, -c(33)] #remove the last column with celltype names
  df <- t(df) #transpose to arrange dataframe for plotting
  df <- data.frame(df)
  df$ST_clusters <- df$X2
  df <- df[, -c(2)]
  
  # replaced ct with 1 because spaces in ct changes the column names
  df[ ,1] <- as.numeric(df[ ,1])
  df$ST_clusters <- factor(df$ST_clusters, levels = unique(df$ST_clusters))
  
  p <- ggplot(df, aes(x = ST_clusters, y = df[,1], fill = ST_clusters)) + 
  geom_bar(position = "dodge", stat = "identity") +
  scale_fill_manual(values = coldef_ST_H) +
  labs(x="ST cluster", y="Expression", title = paste("Celltype proportions of ", ct, " in ST clusters", sep = "")) +
  theme_minimal() +
  theme(plot.title = element_text(hjust = 0.5, face = "bold", size = 12),
        axis.line = element_line(colour = "black"),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        axis.title.x = element_text(size = 10),
        axis.title.y = element_text(size = 10),
        axis.text.x = element_text(size = 8, angle = -45),
        axis.text.y = element_text(size = 8),
        legend.title = element_text(size = 15),
        legend.text = element_text(size = 10))
  
  #ggsave(filename = paste(indir, "/plots/each_celltype-proportions/", ct, ".pdf", sep = ""), plot = p, dpi = 300, height = 12, width = 14)
  
  print(p)
  
}

```

### Expression of predicted celltypes

#### Onto umap

```{r, fig.height=16, fig.width=16, warning=FALSE, verbose=FALSE, message=FALSE}
library(cowplot)
library(ggplot2)
library(gridExtra)

for (type in types){
  cat("For",type, "\n")
  st_data@active.assay = type
  p = lapply(celltypes, function(x) FeaturePlot(st_data, features = x, combine = T, reduction = "umap_after_harmony") + cc + small.leg + NoAxes())
  grid.arrange(grobs = p, ncol=4, top = type)
}
```

